addons/mail/static/src/models/throttle.js

```diff
@@ -0,0 +1,262 @@
+ * cleared:
+ *
+ * - cancel: Canceling a throttle function call means that if a function call is
+ *   pending invocation, cancel removes this pending call invocation. It however
+ *   preserves the internal timer of the cooling down phase of this throttle
+ *   function, meaning that any following throttle function call will be pending
+ *   and has to wait for the remaining time of the cooling down phase before
+ *   being invoked.
+ *
+ * - flush: Flushing a throttle function call means that if a function call is
+ *   pending invocation, flush immediately terminates the cooling down phase and
+ *   the pending function call is immediately invoked. Flush also works without
+ *   any pending function call: it just terminates the cooling down phase, so
+ *   that a following function call is guaranteed to be immediately called.
+ *
+ * - clear: Clearing a throttle function combines canceling and flushing
+ *   together.
+ */
+registerModel({
+    name: 'Throttle',
+    identifyingFields: [[
+        'threadAsThrottleNotifyCurrentPartnerTypingStatus',
+    ]],
+    recordMethods: {
+        /**
+         * Clear any buffered function call and immediately terminates any cooling
+         * down phase in progress.
+         */
+        clear() {
+            if (!this.isCoolingDown) {
+                return;
+            }
+            if (!this.pendingInvokeDeferred) {
+                return;
+            }
+            this.pendingInvokeDeferred.reject(new ThrottleCanceledError(this.id));
+            const coolingDownDeferred = this.coolingDownDeferred;
+            this.update({
+                coolingDownDeferred: clear(),
+                isCoolingDown: false,
+            });
+            coolingDownDeferred.reject(new ThrottleFlushedError(this.id));
+        },
+        /**
+         * 
+         * @param  {...any} args
+         * @returns {any}
+         */
+        async do(...args) {
+            try {
+                // await is important, otherwise errors are not intercepted.
+                return await this._do(...args);
+            } catch (error) {
+                const isSelfReinvokedError = (
+                    error instanceof ThrottleReinvokedError &&
+                    error.throttleId === this.id
+                );
+                const isSelfCanceledError = (
+                    error instanceof ThrottleCanceledError &&
+                    error.throttleId === this.id
+                );
+    
+                if (this.hasSilentCancelationErrors && (isSelfReinvokedError || isSelfCanceledError)) {
+                    // Silently ignore cancelation errors.
+                    // Promise is indefinitely pending for async functions.
+                    return new Promise(() => {});
+                } else {
+                    throw error;
+                }
+            }
+        },
+        onTimeout() {
+            this.coolingDownDeferred.resolve();
+        },
```

Comment:
missing state to know whether we need to invoke func again at the end of cooling down or not?